In [1]:
    !pip install scanpy

     |████████████████████████████████| 2.0 MB 926 kB/s            
  Preparing metadata (setup.py) ... - done
     |████████████████████████████████| 91 kB 7.5 MB/s             
     |████████████████████████████████| 103 kB 56.6 MB/s            
     |████████████████████████████████| 63 kB 1.7 MB/s             
  Created wheel for sinfo: filename=sinfo-0.3.4-py3-none-any.whl size=7899 sha256=65b866f036e156c034f1df052972ef698a823a53d7b4a2fbb0c9f39d4525dd41
  Stored in directory: /root/.cache/pip/wheels/68/ca/56/344d532fe53e855ccd6549795d370588ab8123907eecf4cf30
Successfully built sinfo


![](https://images-wixmp-ed30a86b8c4ca887773594c2.wixmp.com/f/8cc1eeaa-4046-4c4a-ae93-93d656f68688/deztixo-3a7eb036-b7ad-4e91-acf9-29a712e6b94c.jpg?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1cm46YXBwOjdlMGQxODg5ODIyNjQzNzNhNWYwZDQxNWVhMGQyNmUwIiwiaXNzIjoidXJuOmFwcDo3ZTBkMTg4OTgyMjY0MzczYTVmMGQ0MTVlYTBkMjZlMCIsIm9iaiI6W1t7InBhdGgiOiJcL2ZcLzhjYzFlZWFhLTQwNDYtNGM0YS1hZTkzLTkzZDY1NmY2ODY4OFwvZGV6dGl4by0zYTdlYjAzNi1iN2FkLTRlOTEtYWNmOS0yOWE3MTJlNmI5NGMuanBnIn1dXSwiYXVkIjpbInVybjpzZXJ2aWNlOmZpbGUuZG93bmxvYWQiXX0.xsgO6y6Gs6I5f3yDtjuFT-AYcVOmzo0PJ1IbBVonUtw)

### **<span style='color:#F1C40F'>SCE LINKS</span>**
- [Accessing the human atlas](https://bioconductor.org/packages/devel/data/experiment/vignettes/HCAData/inst/doc/hcadata.html)
- [Inerpretability between SCE formats](https://satijalab.org/seurat/archive/v3.0/conversion_vignette.html)
- [Converting SCE b/w R and Python](https://bioconductor.org/packages/devel/bioc/vignettes/zellkonverter/inst/doc/zellkonverter.html)
- [SCE in Python](https://www.kallistobus.tools/tutorials/kb_getting_started/python/kb_intro_2_python/)

# <b>1 <span style='color:#F1C40F'>|</span> BACKGROUND</b>

<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>1.1 | SINGLE CELL EXPERIMENT FORMAT</b></p>
</div>

### **<span style='color:#F1C40F'>ANTIBIOTICS</span>**

- Talk about format

<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>1.2 | COMMON WORKFLOW</b></p>
</div>

- Talk about what steps are usually taken 

<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>1.3 | CUSTOM PYTHON CLASS</b></p>
</div>

In [2]:
import scanpy as sc
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns; sns.set(style='whitegrid')
import plotly.express as px
import plotly.graph_objects as go
import scipy

# Dimensionality Reduction
from sklearn.decomposition import TruncatedSVD
from sklearn.decomposition import PCA
from sklearn.decomposition import SparsePCA
from sklearn.decomposition import KernelPCA
from sklearn.decomposition import IncrementalPCA
from sklearn.decomposition import TruncatedSVD
from sklearn.decomposition import MiniBatchDictionaryLearning
from sklearn.decomposition import FastICA
from sklearn.manifold import Isomap
from sklearn.manifold import MDS
from sklearn.manifold import LocallyLinearEmbedding
from sklearn.manifold import TSNE
from sklearn.random_projection import GaussianRandomProjection
from sklearn.random_projection import SparseRandomProjection
import umap 
from sklearn import metrics

# Clustering
from sklearn.cluster import KMeans
from sklearn.cluster import DBSCAN
from sklearn.cluster import MeanShift
from sklearn.cluster import AgglomerativeClustering
from sklearn.cluster import FeatureAgglomeration

''' CLASS FOR BASIC SCA Related Operations '''

class SCE:
    
    def __init__(self,file=None,title='SCE File',rs = 72):
        self.file = file          # read file string 
        self.title= title         # case name string
        self.assays = {}          # dictionary containing main data
        self.reducedDim = {}      # dictionary containing dim red matrices
        self.reducedDimNames = [] # list containing dimred matrices
        
        self.rs = rs       # random state
        self.dic_dr = None # reserve for dim red dictionary
            
    def info(self):
        print('SCE Instance Information:')
        print(f'Loaded File: {self.file}')
        print(f'Case Name: {self.title}')
        print(f'Input Data Shape: {self.__data.X.shape}')
        print(f'Assays ({len(self.assays)}) : {([ i for i in self.assays.keys()])}')
        print(f'rownames({self.__data.X.shape[0]}) : + \
                         [{self.__data.obs.index[0]} {self.__data.obs.index[1]}' + \
              f' ... {self.__data.obs.index[-1]} ]')
        print(f'colnames({self.__data.X.shape[1]}) : [{self.__data.var.index[1]} +\
                     {self.__data.var.index[1]}' + \
              f' ... {self.__data.var.index[-1]} ]')
        print(f'colData names({len(self.__data.obs.columns.tolist())}) + \
             : {[i for i in self.__data.obs.columns.tolist()]}')
        print(f'reducedDimNames ({len(self.reducedDim.keys())}) : +\
                {([ i for i in self.reducedDim.keys()])}')
        
    ''' READ SCE DATASET '''
        
    def read_h5ad(self):
        self.__data = sc.read_h5ad(self.file)
        
        # For simplicity mouse genes - > human genes 
        genes = [t.upper() for t in self.__data.var.index ]
        self.__data.var.index = genes
        
        self.assays['counts'] = pd.DataFrame(self.__data.X,
                                columns=self.__data.var.index,
                                index=self.__data.obs.index)
        self.obs = self.__data.obs
        
    # expressions per cell
    def epc(self):
        
        fig = plt.figure(figsize = (13,6)); c = 0
        c+=1; fig.add_subplot(1,2,c);
        v = np.asarray(self.assays['counts'].sum(axis = 1)).ravel()
        plt.plot(np.sort(v)) 
        plt.title('Expression per cell')
        plt.xlabel('cells sorted');plt.ylabel('Expression Total')

        c+=1; fig.add_subplot(1,2,c);
        plt.hist(np.sort(v), bins = 100) 
        plt.title('Expression per cell')

        plt.show()
        pd.Series(v).describe()
        
        fig = plt.figure(figsize = (13,6)); c = 0
        c+=1; fig.add_subplot(1,2,c);
        v = np.asarray((self.assays['counts'] != 0).sum(axis = 1)).ravel() 
        plt.plot(np.sort(v)) 
        plt.title('Count genes expressed per cell')
        plt.xlabel('cells sorted')
        plt.ylabel('Count expressed genes')

        c+=1; fig.add_subplot(1,2,c);
        plt.hist(np.sort(v), bins = 100) 
        plt.title('Count genes expressed per cell')

        plt.show()
        pd.Series(v).describe()
    
    ''' LOGCOUNTS '''
    # For scaling of feature data
    
    def logcounts(self,assay='counts'):
        
        if(assay not in self.assays[assay].columns.tolist()):
            print(f'{assay} not found in assays matrix')
            
        # column sums (for each feature/gene)
        libsizes = self.assays[assay].sum(axis=1)
        # scale factor for each col
        sizefactors = libsizes/np.mean(libsizes) 
        # scale data w/ log & store 
        logcounts = np.log2((self.assays[assay].T/sizefactors).T + 1) 
        self.assays['logcounts'] = logcounts
    
    ''' CLUSTERING '''
    # Cluster all cells into groups (nclust)
    def clust_info(self):
        
        print('Hyperparameter Information')
        print('dbscan: [eps][min_samples]')
    
    def clust(self,assay='counts',
                   method='kmeans',
                   nclust=3,  # general
                   eps=1,min_samples=1 # DBSCAN
             ):
    
        X = self.assays[assay]
        
        if(method is 'kmeans'):
            
            print(f'Creating cluster on assay: {assay} using {method}, {nclust} components')
            kmeans_model = KMeans(n_clusters=nclust,
                                  random_state=self.rs).fit(X)

            labels = kmeans_model.labels_
            score = metrics.silhouette_score(X,
                                             labels, 
                                             metric='euclidean')

            print(f'Silhouette Score: {score}')
            self.obs['kmeans'] = labels
            
        # Don't need to specify the number of clusters
        elif(method is 'dbscan'):
            
            print(f'Creating cluster on assay: {assay} using {method}')
            model = DBSCAN(eps=eps,
                           min_samples=min_samples)
            model.fit(X)
            print(model.labels_) # cluster assignments
            self.obs['dbscan'] = model.labels_
            
        elif(method is 'agglomerative'):
            
            print(f'Creating cluster on assay: {assay} using {method}') 
            model = AgglomerativeClustering(n_clusters=nclust)
            model.fit(X)

            # cluster assignments
            print(model.labels_)
            self.obs['agglomerative'] = model.labels_
            
        else:
            
            print('Method not found')
            
#         # Don't need to specift the number of clusters
#         elif(method is 'meanshift'):
            
#             model = MeanShift()
#             model.fit(X)

#             print(f'labels: {model.labels_}') # cluster assignment
#             print(model.cluster_centers_) # cluster centroids
    
    ''' DIMENSIONALITY REDUCTION '''
    # Set Parameter Dictionary for DR
    
    def param_dr(self,method='PCA'):
        
        self.__param_dr_ID = method
        
        if(method is 'PCA'):
            self.dic_dr = {"n_components":2}
        elif(method is 'SPCA'):
            self.dic_dr = {'n_components':2,"alpha":1,
                            "ridge_alpha":0.01,"random_state":self.rs,
                            "max_iter":1000,"n_jobs":-1,
                            "tol":1e-08,"method":"lars"}
        if(method is 'KPCA'):
            self.dic_dr = {'n_components':2,"kernel":"linear",
                        "gamma":None,"degree":3,"coef0":1,
                        "kernel_params":None,"alpha":1.0,
                        "fit_inverse_transform":False,
                        "eigen_solver":"auto","tol":0,
                        "max_iter":None,"iterated_power":'auto',
                        "remove_zero_eig":False,'random_state':self.rs,
                        "n_nobs":-1}
            
        elif(method is 'IPCA'):
            self.dic_dr = {'n_components':2,'whiten':False,'batch_size':None}
        elif(method is 'TSVD'):
            self.dic_dr = {'n_components':2, 'algorithm':'randomized', 
                           'n_iter':5, 'random_state':self.rs,'tol':0.0}
        elif(method is 'GPR'):
            self.dic_dr = {'n_components':'auto','eps':0.1,'random_state':self.rs}
        elif(method is 'SPR'):
            self.dic_dr = {'n_components':'auto', 'density':'auto', 'eps':0.1,
                           'dense_output':False, 'random_state': self.rs}
        elif(method is 'MBD'):
            self.dic_dr = {'n_components':2, 'alpha':1, "n_iter":1000, 
                           "fit_algorithm":'lars', 'n_jobs':-1, 
                           'batch_size':3, 'shuffle':True, 'dict_init':None, 
                           'transform_algorithm':'omp','transform_n_nonzero_coefs':None,
                           'transform_alpha':None, 'verbose':False,'split_sign':False,
                           'random_state':self.rs, 'positive_code':False,
                           'positive_dict':False,'transform_max_iter':1000}
        elif(method is 'ISO'):
            self.dic_dr = {'n_neighbors':5,'n_components':2,'eigen_solver':'auto',
                           'tol':0, 'max_iter':None,'path_method':'auto', 
                           'neighbors_algorithm':'auto', 'n_jobs':self.rs,
                           'metric':'minkowski','p':2, 'metric_params':None}
        elif(method is 'MDS'):
            self.dic_dr = {'n_components':2,'metric':True,"n_init":4, 
                           'max_iter':300,'verbose':0, 'eps':0.001, 'n_jobs':None,
                           'random_state':self.rs,'dissimilarity':'euclidean'}
        elif(method is 'LLE'):
            self.dic_dr = {'n_neighbors':5, 'n_components':2, 'reg':0.001, 
                           'eigen_solver':'auto','tol':1e-06, 'max_iter':100, 
                           'method':'standard', 'hessian_tol':0.0001,
                           'modified_tol':1e-12, 'neighbors_algorithm':'auto', 
                           'random_state':None, 'n_jobs':self.rs}
        elif(method is 'TSNE'):
            self.dic_dr = {'n_components':2,'perplexity':30.0, 
                           'early_exaggeration':12.0,'learning_rate':'warn', 'n_iter':1000, 
                            'n_iter_without_progress':300, 'min_grad_norm':1e-07,
                            'metric':'euclidean', 'init':'warn', 'verbose':0, 
                            'random_state':self.rs, 'method':'barnes_hut', 'angle':0.5,
                            'n_jobs':self.rs, 'square_distances':'legacy'}
        elif(method is 'FICA'):
            self.dic_dr = {'n_components':2,'algorithm':'parallel','whiten':True, 
                        'fun':'logcosh', 'fun_args':None,'max_iter':200, 'tol':0.0001,
                        'w_init':None, 'random_state':self.rs}
            
        print('> DR Parameters Set')
    
    def dr(self,assay='counts',method='PCA',n_comp=2,use_dimRed=None):
            
        # If we already have parameters set
        if(self.dic_dr is not None):
            
            # if set parameter method is not identical
            if(self.__param_dr_ID is not method):
                self.param_dr(method=method)
                self.dic_dr['n_components'] = n_comp # set basic param
                print(f'param_id != dr method, setting default parameters for {method}')   
                
        # use default parameters if nothing is preset
        elif(self.dic_dr is None):
            self.param_dr(method=method)
            self.dic_dr['n_components'] = n_comp # set basic param
    
        # Select Unsupervised Learning Model
        if(method is 'PCA'): model = PCA(**self.dic_dr)
        if(method is 'SPCA'): model = SparsePCA(**self.dic_dr)    
        if(method is 'KPCA'): model = KernelPCA(**self.dic_dr)
        if(method is 'IPCA'): model = IncrementalPCA(**self.dic_dr)
        if(method is 'TSVD'): model = TruncatedSVD(**self.dic_dr)
        if(method is 'GRP'):  model = GaussianRandomProjection(**self.dic_dr)    
        if(method is 'SRP'): model = SparseRandomProjection(**self.dic_dr)
        if(method is 'MBD'): model = MiniBatchDictionaryLearning(**self.dic_dr)
        if(method is 'ISO'): model = Isomap(**self.dic_dr)   
        if(method is 'MDS'): model = MDS(**self.dic_dr)
        if(method is 'LLE'): model = LocallyLinearEmbedding(**self.dic_dr)    
        if(method is 'TSNE'): model = TSNE(**self.dic_dr)
        if(method is 'FICA'): model = FastICA(**self.dic_dr)
        if(method is 'UMAP'): model = umap.UMAP(**self.dic_dr)
        
        if(use_dimRed is None): # Fit on Assay Data
              
            X = model.fit_transform(self.assays[assay])
            self.reducedDim[method] = X # Save Data into 
            print(f'Data saved into reducedDim: {method}')
              
        else: # Fit on already dimensionally reduced data
              
            print(f'Fitting on reducedDim data: {use_dimRed}')
            X = model.fit_transform(self.reducedDim[use_dimRed])
            self.reducedDim[f'{use_dimRed}_{method}'] = X # Save Data into 
            print(f'Data saved into reducedDim: {use_dimRed}_{method}')
            
        if(method is 'PCA'):
            self.__pcavar = np.round(model.explained_variance_ratio_.sum()*100,2)
            print(f'PCA - Total Explained Variance: {self.__pcavar} %')
            self.__pcalabels = {
                str(i): f"PC{i+1} ({var:.1f}%)"
                for i, var in enumerate(model.explained_variance_ratio_ * 100)
            }
            print(self.__pcalabels)

        print('\nParameters used in DR:')
        print(self.dic_dr,'\n')
        
    # Plot Dimensionally Reduced Data
        
    def plot_dr(self,method='PCA',components=[0,1],hue=None):
    
        if(method in self.reducedDim.keys()):

            ttitle = self.title + f' {method}  cells: ' + \
                      str(self.__data.X.shape[0]) + \
                      ' genes: ' + str(self.__data.X.shape[1]) + \
                      ' | geneDim: ' + str(self.reducedDim[method].shape[1]) 

            # Plot Dimensions (1,2)
            if(hue is not None):
                fig = px.scatter(self.reducedDim[method],
                                 x=components[0], y=components[1], 
                                 color=self.obs[hue].values.astype('object'))
#                                  color_discrete_sequence=px.colors.qualitative.T10)
                fig.update_layout(template='plotly_white',height=500,width=600,
                                 font=dict(family='sans-serif',size=12),title=ttitle)
#                 fig.update_traces(marker=dict(size=5.5,opacity=0.75,)) # msize
                fig.update_traces(marker_line=dict(width=1, color='DarkSlateGray'))
                if(method is 'PCA'):
                    fig.update_xaxes(title_text=f'{self.__pcalabels[str(components[0])]}')
                    fig.update_yaxes(title_text=f'{self.__pcalabels[str(components[1])]}')
                else:
                    fig.update_xaxes(title_text=f'{method}{str(components[0])}')
                    fig.update_yaxes(title_text=f'{method}{str(components[1])}')
                fig.show()
            
        else:
            print('Data not found in reducedDim')

<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>1.4 | NORMALISATION</b></p>
</div>

### **REASONS TO USE** **<span style='color:#F1C40F'>NORMALISATION</span>**

- **<span style='color:#F1C40F'>Systematic differences</span>** in sequencing coverage between libraries are often observed in single-cell RNA sequencing data (Stegle, Teichmann, and Marioni 2015)
- Arise from **<span style='color:#F1C40F'>technical differences in cDNA capture</span>** or PCR amplification efficiency across cells, attributable to the difficulty of achieving consistent library preparation with minimal starting material

### **AIM OF** **<span style='color:#F1C40F'>NORMALISATION</span>**

- Normalization aims to remove these differences such that they do not interfere with comparisons of the expression profiles between cells 
- Ensuring that any observed heterogeneity or differential expression within the cell population are driven by biology and not echnical biases

### **SCALING** **<span style='color:#F1C40F'>NORMALISATION</span>**

- This involves dividing all counts for each cell by a **<span style='color:#F1C40F'>cell-specific scaling factor</span>**, often called a “size factor” (Anders and Huber 2010). 
- The assumption here is that any **<span style='color:#F1C40F'>cell-specific bias</span>** (e.g., in capture or amplification efficiency) **<span style='color:#F1C40F'>affects all genes equally via scaling of the expected mean count for that cell</span>**
- The size factor for each cell represents the estimate of the relative bias in that cell, so division of its counts by its size factor should remove that bias
- The resulting “normalized expression values” can then be used for downstream analyses such as **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">dimensionality reduction</mark>** and **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">clustering</mark>**

<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>1.5 | DIMENSIONALITY REDUCTION</b></p>
</div>

- Many scRNA-seq analysis procedures involve: **<span style='color:#F1C40F'>comparing cells based on their expression values across multiple genes</span>**
- For example, **<span style='color:#F1C40F'>clustering aims to identify cells</span>** **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">row data</mark>** with **<span style='color:#F1C40F'>similar transcriptomic profiles</span>** by computing Euclidean distances across genes
- In these applications, **<span style='color:#F1C40F'>each individual gene</span>** **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">row data</mark>** **<span style='color:#F1C40F'>represents a dimension of the data</span>**
    
    
More intuitively:
> - If we had a scRNA-seq data set with two genes, we could make a two-dimensional plot 
> - where each axis represents the expression of one gene and each point in the plot represents a cell



- This concept can be extended to <b>data sets with thousands of genes</b> (columns)
- Where **<span style='color:#F1C40F'>each cell’s expression profile defines its location in the high-dimensional expression space</span>**
- Dimensionality reduction aims to reduce the number of separate dimensions in the data
- **<span style='color:#F1C40F'>This is possible because different genes are correlated if they are affected by the same biological process</span>** 
- Thus, we do not need to store separate information for individual genes, but can instead compress multiple features into a single dimension, e.g., an “eigengene” (Langfelder and Horvath 2007)
> This **reduces computational work** in downstream analyses like clustering, as calculations only need to be performed for a few dimensions **<span style='color:#F1C40F'>rather than thousands of genes</span>**; reduces noise by averaging across multiple genes to obtain a more precise representation of the patterns in the data; and enables effective plotting of the data, for those of us who are not capable of visualizing more than 3 dimensions


<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>1.6 | CLUSTERING</b></p>
</div>

- It is worth stressing the distinction between **<span style='color:#F1C40F'>clusters</span>** and **<span style='color:#F1C40F'>cell</span>** types 
- The **<span style='color:#F1C40F'>former</span>** is an **<span style='color:#F1C40F'>empirical construct</span>** while the latter is a **<span style='color:#F1C40F'>biological truth</span>** (albeit a vaguely defined one)
- For this reason, questions like “what is the true number of clusters?” are usually meaningless. 
- We can **<span style='color:#F1C40F'>define as many clusters as we like</span>**, with **<span style='color:#F1C40F'>whatever algorithm we like</span>**
- Each clustering will represent its own **<span style='color:#F1C40F'>partitioning of the high-dimensional expression space</span>**, and is as “real” as any other clustering.

### **<span style='color:#F1C40F'>CLUSTER APPROXIMATION OF CELL TYPE</span>**

- A more relevant question is **<span style='color:#F1C40F'>“how well do the clusters approximate the cell types?”</span>** 
- Unfortunately, this is difficult to answer given the context-dependent interpretation of biological truth
- Some analysts will be satisfied with resolution of the major cell types; other analysts may want resolution of subtypes; and others still may require resolution of different states (e.g., metabolic activity, stress) within those subtypes
- Moreover, two clusterings can be highly inconsistent yet both valid, simply partitioning the cells based on different aspects of biology. 
- Indeed, asking for an unqualified “best” clustering is akin to asking for the best magnification on a microscope without any context

### **<span style='color:#F1C40F'>CLUSTER AS A TOOL</span>**



- It is helpful to realize that clustering, like a microscope, is simply a tool to explore the data
- We can zoom in and out by **<span style='color:#F1C40F'>changing the resolution of the clustering parameters</span>**, and we can experiment with different **<span style='color:#F1C40F'>clustering algorithms</span>**
- Aiming to obtain **<span style='color:#F1C40F'>alternative perspectives</span>** of the data
- This iterative approach is entirely permissible for data exploration, which constitutes the majority of all scRNA-seq data analyses

# <b>2 <span style='color:#F1C40F'>|</span> SCE CLASS DATA</b>

<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>2.1 | LOADING PRELOADED SCE</b></p>
</div>

- Two publically available SCE datasets were extracted from **<span style='color:#F1C40F'>Bioconductor</span>** & saved into the **[Bioinformatics dataset](https://www.kaggle.com/shtrausslearning/bioinformatics)**
- Aside from **<span style='color:#F1C40F'>loom</span>**, **<span style='color:#F1C40F'>Seurat</span>**, **<span style='color:#F1C40F'>anndata</span>**, we can save **<span style='color:#F1C40F'>SCE</span>** data in standard **<span style='color:#F1C40F'>H5AD</span>** format
- Let's load the mouse brain single-cell RNA-seq data from the **[paper](https://www.science.org/doi/10.1126/science.aaa1934?url_ver=Z39.88-2003&rfr_id=ori:rid:crossref.org&rfr_dat=cr_pub%20%200pubmed)**

In [3]:
zeisel = SCE('/kaggle/input/bioinformatics/SCE/zeisel.h5ad') # initialise new instance of class SCE
zeisel.read_h5ad() # read the h5ad data

<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>2.2 | SHOWING SCE INFORMATION</b></p>
</div>

- Display the **<span style='color:#F1C40F'>class information</span>** using **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">.info()</mark>**

In [4]:
zeisel.info()

SCE Instance Information:
Loaded File: /kaggle/input/bioinformatics/SCE/zeisel.h5ad
Case Name: SCE File
Input Data Shape: (3005, 20006)
Assays (1) : ['counts']
rownames(3005) : +                          [1772071015_C02 1772071017_G12 ... 1772058148_F03 ]
colnames(20006) : [TSHZ1 +                     TSHZ1 ... MT-ND4L ]
colData names(10) +              : ['tissue', 'group #', 'total mRNA mol', 'well', 'sex', 'age', 'diameter', 'cell_id', 'level1class', 'level2class']
reducedDimNames (0) : +                []


<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>2.3 | NORMALISATION</b></p>
</div>

In [5]:
zeisel.logcounts()
zeisel.info()

counts not found in assays matrix
SCE Instance Information:
Loaded File: /kaggle/input/bioinformatics/SCE/zeisel.h5ad
Case Name: SCE File
Input Data Shape: (3005, 20006)
Assays (2) : ['counts', 'logcounts']
rownames(3005) : +                          [1772071015_C02 1772071017_G12 ... 1772058148_F03 ]
colnames(20006) : [TSHZ1 +                     TSHZ1 ... MT-ND4L ]
colData names(10) +              : ['tissue', 'group #', 'total mRNA mol', 'well', 'sex', 'age', 'diameter', 'cell_id', 'level1class', 'level2class']
reducedDimNames (0) : +                []


In [6]:
zeisel.assays['logcounts'].head()

TSPAN12     TSHZ1    FNBP1L  ADAMTS15    CLDN12  RXFP1  \
1772071015_C02  0.000000  1.588285  1.588285       0.0  0.738960    0.0   
1772071017_G12  0.000000  0.725879  0.725879       0.0  0.725879    0.0   
1772071017_A05  0.000000  0.000000  1.907973       0.0  0.544785    0.0   
1772071014_B06  1.225389  0.920006  1.477287       0.0  0.000000    0.0   
1772067065_H06  0.000000  1.250143  0.756447       0.0  0.000000    0.0   

                2310042E22RIK    SEMA3C      JAM2  APBB1IP  ...    MT-ND5  \
1772071015_C02       0.000000  3.063278  0.738960      0.0  ...  3.174354   
1772071017_G12       0.000000  0.000000  0.000000      0.0  ...  2.783492   
1772071017_A05       0.939306  3.640397  0.544785      0.0  ...  2.978569   
1772071014_B06       1.225389  0.532129  0.000000      0.0  ...  2.326121   
1772067065_H06       0.000000  2.980619  0.000000      0.0  ...  1.617286   

                  MT-ND1    MT-CO3   MT-CYTB   MT-ATP8    MT-CO2    MT-CO1  \
1772071015_C02  5.732808  5.517942  4.794878  5.871145  5.475189  5.677309   
1772071017_G12  5.718600  5.965030  4.895881  4.863841  6.009639  6.201315   
1772071017_A05  5.008925  5.564959  4.857332  4.239803  4.924260  5.478527   
1772071014_B06  5.106625  4.596649  4.569803  5.720912  4.906602  3.890243   
1772067065_H06  3.101410  3.745006  2.703676  2.703676  3.212865  4.391659   

                 MT-RNR2   MT-RNR1   MT-ND4L  
1772071015_C02  6.737385  4.442157  3.277485  
1772071017_G12  7.742639  5.871446  3.344077  
1772071017_A05  6.598272  5.507917  2.358740  
1772071014_B06  5.106625  3.244152  2.667379  
1772067065_H06  4.438276  0.756447  2.542326  

[5 rows x 20006 columns]

<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>2.4 | DIMENSIONALITY REDUCTION</b></p>
</div>

Using **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">group #</mark>** colour identification, let's look at three cases:

- **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">PCA</mark>**, **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">counts</mark>** assays data, reducing the number of genes from 20006 to 2 components, plotting all 3005 cell data
- **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">PCA</mark>**, & normalised **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">logcounts</mark>** assays data, reducing the number of genes from 20006 to 2 components, plotting all 3005 cell data
- **<mark style="background-color:#F1C40F;color:white;border-radius:5px;opacity:0.9">UMAP</mark>** manifold dimensionality reduction technique (**[reference](https://arxiv.org/abs/1802.03426)**), reducing the number of genes from 20006 to 2 components, plotting all 3005 cell data

In [7]:
zeisel.dr(method='PCA',n_comp=2)
zeisel.plot_dr(method='PCA',hue='group #')

> DR Parameters Set
Data saved into reducedDim: PCA
PCA - Total Explained Variance: 60.26 %
{'0': 'PC1 (49.6%)', '1': 'PC2 (10.7%)'}

Parameters used in DR:
{'n_components': 2} 



In [8]:
zeisel.dr(assay='logcounts',method='PCA',n_comp=2)
zeisel.plot_dr(method='PCA',hue='group #')

Data saved into reducedDim: PCA
PCA - Total Explained Variance: 18.62 %
{'0': 'PC1 (14.6%)', '1': 'PC2 (4.0%)'}

Parameters used in DR:
{'n_components': 2} 



In [9]:
zeisel.dr(method='UMAP',n_comp=2)
zeisel.plot_dr(method='UMAP',hue='group #')

> DR Parameters Set
param_id != dr method, setting default parameters for UMAP
Data saved into reducedDim: UMAP

Parameters used in DR:
{'n_components': 2} 



<div style="color:white;display:fill;border-radius:8px;
            background-color:#323232;font-size:150%;
            font-family:Nexa;letter-spacing:0.5px">
    <p style="padding: 8px;color:white;"><b>2.4 | CLUSTERING</b></p>
</div>



In [10]:
zeisel.clust(assay='counts',method='kmeans',nclust=9)
zeisel.plot_dr(method='UMAP',hue='kmeans')

Creating cluster on assay: counts using kmeans, 9 components
Silhouette Score: 0.25307774543762207


In [11]:
zeisel.clust(assay='counts',method='agglomerative',nclust=3)
zeisel.plot_dr(method='UMAP',hue='agglomerative')

Creating cluster on assay: counts using agglomerative
[2 2 2 ... 2 2 2]


In [12]:
zeisel.clust_info()

Hyperparameter Information
dbscan: [eps][min_samples]


In [13]:
zeisel.clust(assay='counts',method='dbscan',eps=0.01,min_samples=2)
zeisel.plot_dr(method='UMAP',hue='dbscan')

Creating cluster on assay: counts using dbscan
[-1 -1 -1 ... -1 -1 -1]
